ta/pkcs11/src/pkcs11_token.c

```diff
@@ -111,3 +111,54 @@ uint32_t entry_ck_slot_list(uint32_t ptypes, TEE_Param *params)
 
 	return PKCS11_CKR_OK;
 }
+
+uint32_t entry_ck_slot_info(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = &params[0];
+	TEE_Param *out = &params[2];
+	uint32_t rv = 0;
+	struct serialargs ctrlargs = { };
+	uint32_t token_id = 0;
+	struct ck_token *token = NULL;
+	const char desc[] = PKCS11_SLOT_DESCRIPTION;
+	const char manuf[] = PKCS11_SLOT_MANUFACTURER;
+	const char hwver[2] = PKCS11_SLOT_HW_VERSION;
+	const char fwver[2] = PKCS11_SLOT_FW_VERSION;
+	struct pkcs11_slot_info info = { };
+
+	if (ptypes != exp_pt ||
+	    out->memref.size != sizeof(struct pkcs11_slot_info))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
```

Comment:
This seems like a complicated version of:
``` C
if (ctrl->memref.size != sizeof(token_id)
    return something_bad;
memcpy(&token_id, ctrl->memref.buffer, sizeof(token_id));
```

Reply:
Almost all TA command handler need to extract data serialized inside a buffer provided by client.
I really prefer to use friendly helper functions to factorize the extraction and related checks.

In this function `entry_ck_slot_info()`, it is quite simple as there is only 1 32bit data to extract, but other commands will provide more complex data, i.e. the command related to [`C_SetPin()`](https://github.com/OP-TEE/optee_os/pull/2732/files#diff-9b89a766b9b62bace834e6ebd6823ea9R158) or even more complex [object attribute templates](https://github.com/OP-TEE/optee_os/pull/2732/files#diff-9b89a766b9b62bace834e6ebd6823ea9R281).

Reply:
To me it doesn't make much sense, but it's not too horrible either.  If you think it's important keep it.
I'm looking forward to see how the object attributes are serialized with this.

Reply:
[Here](https://github.com/etienne-lms/optee_os/blob/sks/ta/sks/src/processing.c#L331) and [here](https://github.com/etienne-lms/optee_os/blob/sks/ta/sks/src/processing.c#L903) are examples (from my github tree) of de-serailization of client input paremeters.